Support opensearch-go tracing

keisku opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Commit and PR titles should be prefixed with the general area of the pull request's change.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

Support [opensearch-project/opensearch-go](https://github.com/opensearch-project/opensearch-go) by manual / orchestrion instrumentation.
Tested OpenSearch v2 only because of [Release Schedule and Maintenance Policy](https://opensearch.org/releases.html).
- The maintenance window of OpenSearch v1 will end April 29th, 2025.
- OpenSearch v3 is still not GA.

`go get -u github.com/aws/aws-sdk-go-v2/...` fixes the issue described https://github.com/aws/aws-sdk-go-v2/issues/2370#issuecomment-1953308268. 
It's a bit complicated, I confirmed this issue in the test code when `go get github.com/opensearch-project/opensearch-go/v4` indirectly updated `github.com/aws/aws-sdk-go`.

https://github.com/DataDog/dd-trace-go/blob/25864e0e88c85b117866e502594d639de41b4d08/contrib/aws/aws-sdk-go-v2/aws/aws_test.go#L108-L115

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
* If this resolves a GitHub issue, include "Fixes #XXXX" to link the issue and auto-close it on merge.
-->

https://github.com/DataDog/dd-trace-go/issues/2186

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).
-->

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.
- [ ] For internal contributors, a matching PR should be created to the `v2-dev` branch and reviewed by @DataDog/apm-go.


Unsure? Have a question? Request a review!

rarguelloF commented on contrib/opensearch-project/opensearch-go.v4/opensearch.go:
```diff
@@ -0,0 +1,222 @@
+}
+
+// WrapRoundTripper wraps an http.RoundTripper with tracing.
+func WrapRoundTripper(rt http.RoundTripper) http.RoundTripper {
```
```suggestion
func WrapRoundTripper(rt http.RoundTripper, opts ...Option) http.RoundTripper {
```

rarguelloF commented on contrib/opensearch-project/opensearch-go.v4/opensearch.go:
```diff
@@ -0,0 +1,222 @@
+}
+
+// WrapClient wraps OpenSearch client with tracing.
+func WrapClient(c *opensearch.Client, opts ...Option) {
```
```suggestion
func TraceClient(c *opensearch.Client, opts ...Option) {
```

in the context of this repo, wrapping usually means the function would return another `*opensearch.Client`, and this function modifies the received one.

rarguelloF commented on contrib/opensearch-project/opensearch-go.v4/opensearch.go:
```diff
@@ -0,0 +1,222 @@
+
+// NewClient returns a new opensearch.Client enhanced with tracing.
+func NewClient(cfg opensearch.Config, opts ...Option) (*opensearch.Client, error) {
+	if cfg.Transport == nil {
+		cfg.Transport = WrapRoundTripper(http.DefaultTransport)
+	} else {
+		cfg.Transport = WrapRoundTripper(cfg.Transport)
+	}
+	c, err := opensearch.NewClient(cfg)
+	if err != nil {
+		return nil, err
+	}
+	tracerConfig := defaultConfig()
+	for _, fn := range opts {
+		fn(tracerConfig)
+	}
+	t := &transport{
+		opensearchtransport: c.Transport,
+		config:              tracerConfig,
+	}
+	c.Transport = t
+	return c, nil
```
What is the difference / advantage of tracing both the transport and the roundtripper? Would it be possible to achieve the same functionality using one of the 2 only?

keisku commented:
@rarguelloF Thank you for reviewing this PR.
I am going to close this PR since this feature is not going to be released as v1.

I created a new PR for `v2-dev`, https://github.com/DataDog/dd-trace-go/pull/3280, after addressing your suggestions in this PR!
